src/nav/InitialNavigationDispatcher.js

```diff
@@ -0,0 +1,83 @@
+/* @flow strict-local */
+import type { Node as React$Node } from 'react';
+import { PureComponent } from 'react';
+
+import type { Dispatch, Account, User } from '../types';
+import { resetToAccountPicker, resetToRealmScreen, resetToMainTabs } from '../actions';
+import { connect } from '../react-redux';
+import { getIsHydrated, hasAuth as getHasAuth } from '../selectors';
+
+type SelectorProps = $ReadOnly<{|
+  isHydrated: boolean,
+  hasAuth: boolean,
+  accounts: Account[],
+  users: User[],
+|}>;
+
+type Props = $ReadOnly<{|
+  children: React$Node,
+
+  dispatch: Dispatch,
+  ...SelectorProps,
+|}>;
+
+class InitialNavigationDispatcher extends PureComponent<Props> {
+  componentDidMount() {
+    if (this.props.isHydrated) {
+      this.doInitialNavigation();
+    }
+  }
+
+  componentDidUpdate(prevProps) {
+    if (!prevProps.isHydrated && this.props.isHydrated) {
+      this.doInitialNavigation();
+    }
+  }
+
+  /**
+   * Data has been loaded, so open the app to the right screen.
+   *
+   * Not to be called before the REHYDRATE action, and not to be
+   * called more than once.
+   */
+  doInitialNavigation = () => {
+    const { hasAuth, accounts, users, dispatch } = this.props;
+
+    // If there are accounts but the active account is not logged in,
+    // show account screen.
+    if (!hasAuth) {
+      if (accounts.length > 1) {
+        dispatch(resetToAccountPicker());
+        return;
+      } else {
+        dispatch(resetToRealmScreen({ initial: true }));
+        return;
+      }
+    }
+
+    // If there's an active, logged-in account but no server data, then behave
+    // like `ACCOUNT_SWITCH`: show loading screen.  Crucially, `sessionReducer`
+    // will have set `needsInitialFetch`, too, so we really will be loading.
+    //
+    // (Valid server data must have a user: the self user, at a minimum.)
+    if (users.length === 0) {
+      // We're already on the loading screen.
```

Comment:
This calls for a few words about why we believe that's the case. (I think these lines originated with https://github.com/zulip/zulip-mobile/commit/57f1a7630f9d0b69f15349fa6ab61afd81242dac in your first revision of this branch, where this code is in a method on `LoadingScreen` and so the context already makes that clear.)